src/rt/cpp/when.h

```diff
@@ -55,24 +78,30 @@ namespace verona::cpp
       else
       {
         auto p = std::get<index>(cown_tuple);
-        array[index] = p;
-        assert(array[index] != nullptr);
-        array_assign<index + 1>(array);
+        if constexpr (is_read_only<decltype(p)>())
+          requests[index] = Request::read(p.t);
+        else
+          requests[index] = Request::write(p.t);
+        assert(requests[index].cown() != nullptr);
+        array_assign<index + 1>(requests);
       }
     }
 
-    When(ActualCown<Args>*... args) : cown_tuple(args...) {}
+    When(Access<Args>... args) : cown_tuple(args...) {}
 
     /**
      * Converts a single `cown_ptr` into a `acquired_cown`.
      *
      * Needs to be a separate function for the template parameter to work.
      */
     template<typename C>
-    static acquired_cown<C> actual_cown_to_acquired(ActualCown<C>* c)
+    static acquired_cown<C> access_to_acquired(Access<C> c)
     {
-      assert(c != nullptr);
-      return acquired_cown<C>(*c);
+      assert(c.t != nullptr);
+      if constexpr (std::is_const<C>())
+        return acquired_cown<C>(*reinterpret_cast<ActualCown<const C>*>(c.t));

```

Comment:
Could the acquired_cown constructor be improved to not need the reinterpret_cast? Or a private conversion in ActualCown. Intuition says that adding const should not require reinterpret-cast.

Reply:
I think I have done what you are suggesting in:
[d3fd1a0](https://github.com/microsoft/verona/pull/595/commits/d3fd1a0c97b377e6475b94220097517320e06d5c)

Reply:
Excellent!